### PR TITLE
Add trigger contracts

### PR DESCRIPTION
**Pull request by mds1**
Ported from https://github.com/Cozy-Finance/cozy-protocol-v2/tree/4ff022038741130d37efd1e8ce10f8b8a7e0114d

Changes from that commit:
- Use of `makePersistent` cheatcode due to a change in how foundry fork storage works. Previously, your own code/contracts would exist on all forks at the same address. The change resulted in your own code/contracts being tied to a fork (which makes sense) and the use `vm.makePersistent` to tell foundry you want that code at all addresses. This is also why the UMA tests use `createSelectFork` instead of `createFork` now (slightly different use case, where we want the contract deployed on the fork in `setUp` but it shouldn't be persistent across forks in this case)
- More unit test-y, i.e. removal of reliance on `manager.createSet` when testing `BaseTrigger`, etc. A future PR

Suggestions for cleanup of this repo in future PRs:
1. I think we may be able to remove reliance on `MarketInfo` struct and possibly some other related interfaces
2. Remove `MockManager` and `MockChainlinkTrigger` contracts and use `vm.mockCall` instead
3. Pin optimism tests to fork block (already being tracked)
4. Add deploy scripts
5. Clean up tests into a more consistent pattern, such as the one where each `contract` unit tests a single function, e.g. `contract RunProgrammaticCheck is TriggerTestSetup { ... }`, then integration tests go in e.g. `contract Integration is TriggerTestSetup { ... }

### CODE REVIEW

File: script/Contract.s.sol [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.13;
+
+import "forge-std/Script.sol";
+
+contract ContractScript is Script {
```
Comment (chad-js) [inline script/Contract.s.sol:6]: can we remove this?

File: src/interfaces/IManager.sol [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+pragma solidity 0.8.15;
+
+import "src/interfaces/ICState.sol";
+import "src/interfaces/ISet.sol";
+
+/**
+ * @dev Interface for interacting with the Cozy protocol Manager. This is not a comprehensive
+ * interface, and only contains the methods needed by triggers.
+ */
+interface IManager is ICState {
```
Comment (chad-js) [inline src/interfaces/IManager.sol:11]: if we import `cozy-v2-triggers` as a dependency in `cozy-v2-developer-guides`, how should we go about extending this `IManager` there - would there be a naming conflict if it's also called `IManager`? For example, in that repo we need `createSet` https://github.com/Cozy-Finance/cozy-v2-developer-guides/blob/main/src/interfaces/IManager.sol#L12. I guess we need to name it differently there and inherit this `IManager`?

File: script/Contract.s.sol (hunk 1 above)
Comment (mds1) [reply script/Contract.s.sol:6]: I'm pretty sure git doesn't let you commit empty folders which is why I kept it for now

File: test/utils/TriggerTestSetup.sol [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+  function setUp() public virtual {
+    // Create addresses.
+    asset = IERC20(makeAddr("asset"));
+    costModel = ICostModel(makeAddr("costModel"));
+    localOwner = makeAddr("localOwner");
+    localPauser = makeAddr("localPauser");
+    manager = IManager(makeAddr("manager"));
+    set = ISet(makeAddr("set"));
+    set2 = ISet(makeAddr("set2"));
```
Comment (chad-js) [inline test/utils/TriggerTestSetup.sol:35]: ooo `makeAddr`, nice

File: script/Contract.s.sol (hunk 1 above)
Comment (chad-js) [reply script/Contract.s.sol:6]: oh lol okay 😄

File: src/interfaces/IManager.sol (hunk 2 above)
Comment (mds1) [reply src/interfaces/IManager.sol:11]: with `as` syntax! (we probably could have used to get our current cozy-protocol-v2 compiling, I don't recall exactly where all the conflicts were)

```solidity
// file: IManager.sol in cozy-v2-developer-guides repo
import { IManager as IManagerBase } from "cozy-v2-triggers/interfaces/IManager.sol";

interface IManager is IManagerBase {
  // --- snip ---
}
```

Comment (chad-js) [reply src/interfaces/IManager.sol:11]: ooooh perfect, thanks. leaving unresolved for david to 👁️

File: test/utils/TriggerTestSetup.sol (hunk 3 above)
Comment (mds1) [reply test/utils/TriggerTestSetup.sol:35]: Yea new util in forge-std, I like it! https://github.com/foundry-rs/forge-std/pull/146

**Review by chad-js — APPROVED** [review]
thanks matt!

File: test/utils/TriggerTestSetup.sol (hunk 3 above)
Comment (chad-js) [reply test/utils/TriggerTestSetup.sol:35]: sweet, leaving open for david to 👁️

Comment (davidlaprade) [reply test/utils/TriggerTestSetup.sol:35]: Oh, that's really nice!

**Review by davidlaprade — APPROVED** [review]
Looks great, thanks!

**Pull request merged**